### PR TITLE
Rework how the junit xml-style log file is generated

### PR DESCRIPTION
**Pull request by shs96c**

### CODE REVIEW

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOutputListener.java [hunk 1]
```diff
@@ -45,67 +43,125 @@ public BazelJUnitOutputListener(Path xmlOut) {
...
+      for (Map.Entry<TestData, List<TestData>> suiteAndTests : testSuites.entrySet()) {
+        new TestSuiteXmlRenderer(testPlan)
+            .toXml(xml, suiteAndTests.getKey(), suiteAndTests.getValue());
+      }
+    } catch (XMLStreamException e) {
+      throw new RuntimeException(e);
+    }
+
+    // Delete the results we've used to conserve memory
+    Stream.concat(testCases.stream(), testSuites.keySet().stream())
```
Comment (illicitonion) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOutputListener.java:146]: What guarantees do we have that the suites are complete at this point and won't have more child test-cases which reference them? Given the code in `matchTestCasesToSuites` ignores any tests where we don't know about their containing suite, I'm quite wary of throwing away test suites unless we're very sure we're done with them (which could do with a comment justifying why we we're so sure :))

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md [hunk 2]
```diff
@@ -0,0 +1,280 @@
...
+those `testsuite` entries is a `testcase` per method in the test class, 
+where each `testcase` has a `name` attribute that is a combination of the 
+test method name without signature information and the run number (eg. 
+`someTest[1]`)
+
+#### Pathological
+
+Skipped
+
+## Desired behaviour
```
Comment (illicitonion) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md:221]: Given all of the (wonderful!) effort that went into producing this spec, would it be possible to check in a golden XML file for the comparative tests, run the test, and assert that the XML matches? (even if we need to be a bit fuzzy zeroing-out timestamps and stuff)

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md [hunk 3]
```diff
@@ -0,0 +1,280 @@
...
+JUnit5 has a number of different ways of running and not running tests, so a 
+quick review of how each of the existing test frameworks. For each of these 
+we'll consider the various following combinations:
+
+* "Simple": A regular test case, using no special features
+* "Suite": A test suite, created using the `@Suite` annotation and containing 
+  two test classes
+* "Dynamic" A dynamic test using the `@TestFactory` annotation
+* "Parameterised" A parameterised test case
+* "Pathological" A parameterised test case where the test method is disabled
```
Comment (illicitonion) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md:32]: Could you rename `DisabledSkippedParameterisedTest` in the comparative-tests to `PathologicalTest`, or rename it in this file to `DisabledSkippedParameterisedTest`? All of the others line up nicely, but I had t odo some reading around to understand what example test this was referring to.

File: comparative-tests/src/test/java/com/github/bazel_contrib/contrib_rules_jvm/comparative/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,22 @@
+# gazelle:java_extension disabled
```
Comment (illicitonion) [inline comparative-tests/src/test/java/com/github/bazel_contrib/contrib_rules_jvm/comparative/BUILD.bazel:1]: I'd be tempted to add checked in example output XMLs from each of the test-runners, so someone can inspect them at a glance without having to go through all the hoops of having to run them themselves.

**Review by illicitonion — COMMENTED** [review]
Thanks for all of the effort into doing the comparisons, it's really useful!

Generally looking good, a few questions

File: comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.PathologicalTest.xml [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    <property name="java.vendor" value="Apple Inc."/>
+    <property name="java.vm.version" value="17.0.8+7-LTS"/>
+    <property name="sun.io.unicode.encoding" value="UnicodeBig"/>
+    <property name="java.class.version" value="61.0"/>
+  </properties>
+  <testcase name="bootstrap(String)" classname="com.github.bazel_contrib.contrib_rules_jvm.comparative.PathologicalTest" time="0.0">
+    <skipped message="Environment variable [PATH] with value [/Users/shs/.nvm/versions/node/v16.17.0/bin:/usr/local/bin:/System/Cryptexes/App/usr/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/homebrew/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/local/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/bin:/var/run/com.apple.security.cryptexd/codex.system/bootstrap/usr/appleinternal/bin:/opt/appleinternal/bin:/opt/apple/bin:/Library/Apple/usr/bin:/opt/xbs/bin:/usr/appleinternal/bin:/private/var/tanuki:/usr/local/share/dotnet:~/.dotnet/tools:/Users/shs/.cargo/bin:/Users/shs/Library/Application Support/JetBrains/Toolbox/scripts:/opt/homebrew/bin:/Users/shs/bin:/Users/shs/.adt/bin:/Users/shs/.dotnet/tools] matches regular expression [.*]"/>
```
Comment (illicitonion) [inline comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.PathologicalTest.xml:56]: You may want to remove a few entries from this output

**Review by illicitonion — APPROVED** [review]
Thanks!

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOutputListener.java [hunk 6]
```diff
@@ -45,61 +46,127 @@ public BazelJUnitOutputListener(Path xmlOut) {
...
-    }
+    this.testPlan = testPlan;
 
-    roots =
-        testPlan.getRoots().stream()
-            .map(root -> new RootContainer(root, testPlan))
-            .collect(
-                collectingAndThen(toCollection(LinkedHashSet::new), Collections::unmodifiableSet));
+    try {
+      // Closed when we call `testPlanExecutionFinished
```
Comment (blorente) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOutputListener.java:52]:
```suggestion
      // Closed when we call `testPlanExecutionFinished`
```

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md [hunk 7]
```diff
@@ -0,0 +1,280 @@
...
+
+#### Simple
+
+Below the `testsuites` tag is a single `testsuite` tag. This is named after 
+the test class name. Nested below this is a `testcase` for each test method. 
+The `name` attribute is the name of the test method without type information 
+(eg. `example`)
+
+#### Suite
+
```
Comment (blorente) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md:195]:
```suggestion

Skipped

```

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md [hunk 8]
```diff
@@ -0,0 +1,280 @@
...
+is available to it. These are arranged in a hierarchy, with "test roots" 
+being the actual engine being used. These test roots are used by the legacy 
+XML listener to provide the suite name, which is why they bear no relation 
+to a class name at all. 
+
+Neither Gradle nor Maven expose the root containers, so it's appropriate 
+that we don't do that either.
+
+The reason that the legacy XML listener is outputting top-level test suites 
+with names like `JUnit Platform Suite` is that
```
Comment (blorente) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md:239]: What is the reason? *from the edge of seat* I gotta know! (probably something to do with JUnit 4 compatibility?)

File: java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md [hunk 9]
```diff
@@ -0,0 +1,280 @@
+# JUnit 5 Test Runner
```
Comment (blorente) [inline java/src/com/github/bazel_contrib/contrib_rules_jvm/junit5/README.md:1]: This document is amazing and it should be shortlinked from the documentations of most of the tools mentioned here. Thanks for the effort writing it!

File: comparative-tests/src/test/java/com/github/bazel_contrib/contrib_rules_jvm/comparative/BUILD.bazel (hunk 4 above)
Comment (blorente) [reply comparative-tests/src/test/java/com/github/bazel_contrib/contrib_rules_jvm/comparative/BUILD.bazel:1]: Thanks for doing it, it's great to be able to inspect them as I read the fantastic Readme.

File: comparative-tests/gradlew.bat [hunk 10]
```diff
@@ -0,0 +1,92 @@
+@rem
```
Comment (blorente) [inline comparative-tests/gradlew.bat:1]: Gotta love them gradle wrappers.

File: comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.DynamicTests.xml [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    <property name="java.runtime.version" value="17.0.8+7-LTS"/>
+    <property name="user.name" value="shs"/>
+    <property name="path.separator" value=":"/>
+    <property name="os.version" value="14.2"/>
+    <property name="java.runtime.name" value="OpenJDK Runtime Environment"/>
+    <property name="file.encoding" value="UTF-8"/>
+    <property name="java.vm.name" value="OpenJDK 64-Bit Server VM"/>
+    <property name="java.vendor.version" value="AppleJDK-17.0.8.7.2"/>
+    <property name="localRepository" value="/Users/shs/.m2/repository"/>
+    <property name="java.vendor.url.bug" value="https://rdar.apple.com/new/1003593"/>
```
Comment (blorente) [inline comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.DynamicTests.xml:41]:
```suggestion
```

**Review by blorente — APPROVED** [review]

File: comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.PathologicalTest.xml (hunk 5 above)
Comment (shs96c) [reply comparative-tests/target/surefire-reports/TEST-com.github.bazel_contrib.contrib_rules_jvm.comparative.PathologicalTest.xml]: Done.

File: java/test/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOuputListenerTest.java [hunk 12]
```diff
@@ -231,11 +218,22 @@ public void ensureOutputsAreProperlyEscaped() {
...
+  }
+
+  private Document generateDocument(XmlGenerator renderer) {
     try {
       Writer writer = new StringWriter();
       XMLStreamWriter xsw = XMLOutputFactory.newDefaultFactory().createXMLStreamWriter(writer);
-      result.toXml(xsw);
+
+      renderer.accept(xsw);
+      System.out.println(writer);
```
Comment (rdesgroppes) [inline java/test/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOuputListenerTest.java:236]: leftover?

Comment (shs96c) [reply java/test/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOuputListenerTest.java:236]: Oops. Yes. Leftover.

Comment (shs96c) [reply java/test/com/github/bazel_contrib/contrib_rules_jvm/junit5/BazelJUnitOuputListenerTest.java:236]: Created #228 to fix this

**Pull request merged**